plugins/mediatek-scaler/fu-mediatek-scaler-device.c

```diff
@@ -0,0 +1,1063 @@
+	g_byte_array_append(ddc_msgbox_write, ddc_wfmt, sizeof(ddc_wfmt));
+	g_byte_array_append(ddc_msgbox_write, buf_in->data, buf_in->len);
+
+	chksum ^= DDC_DISPLAY_I2C_ADDR;
+	for (gsize i = 0; i < ddc_msgbox_write->len; i++)
+		chksum ^= ddc_msgbox_write->data[i];
+	g_byte_array_append(ddc_msgbox_write, &chksum, 1);
+
+	/* print the raw data */
+	rawbytes = g_bytes_new(ddc_msgbox_write->data, ddc_msgbox_write->len);
+	fu_dump_bytes(G_LOG_DOMAIN, "DDC/CI write message", rawbytes);
+
+	return fu_udev_device_pwrite(FU_UDEV_DEVICE(self->i2c_dev),
+				     0x0,
+				     ddc_msgbox_write->data,
+				     ddc_msgbox_write->len,
+				     error);
+}
+
+static gboolean
+mediatek_scaler_device_ddc_read(FuMediatekScalerDevice *self,
+				GByteArray *buf_in,
+				GByteArray *buf_out,
+				GError **error)
+{
+	guint8 buf[0x40] = {0x00}; /* default 64 bytes */
+	g_autoptr(GBytes) rawbytes = NULL;
+	gsize report_data_sz = 0;
+	guint8 checksum = 0;
+
+	/* write for read */
+	if (!mediatek_scaler_device_ddc_write(self, buf_in, error))
+		return FALSE;
+
+	/* DDCCI spec requires host to wait at least 50 - 200ms before next message */
+	fu_device_sleep(FU_DEVICE(self), FU_MEDIATEK_SCALER_DDC_MSG_DELAY_MS);
+
+	/* read into tmp buffer */
+	if (!fu_udev_device_pread(FU_UDEV_DEVICE(self->i2c_dev), 0x0, buf, sizeof(buf), error))
+		return FALSE;
+
+	/* read buffer = addr(src) + length + data + checksum */
+	rawbytes = g_bytes_new(buf, sizeof(buf));
+	fu_dump_bytes(G_LOG_DOMAIN, "DDC/CI read buffer", rawbytes);
+
+	/* verify read buffer: [0] == source address */
+	if (buf[0] != DDC_DISPLAY_I2C_ADDR) {
+		g_set_error(error,
+			    G_IO_ERROR,
+			    G_IO_ERROR_INVALID_DATA,
+			    "invalid read buffer: addr(src) expected 0x%02x, got 0x%02x.",
+			    (guint)DDC_DISPLAY_I2C_ADDR,
+			    buf[0]);
+		return FALSE;
+	}
+
+	/* verify read buffer: [1] as the length of data */
+	if (buf[1] <= DDC_DATA_LEN_DFT) {
+		g_set_error(error,
+			    G_IO_ERROR,
+			    G_IO_ERROR_INVALID_DATA,
+			    "invalid read buffer: size 0x%02x must greater than 0x%02x.",
+			    buf[1],
+			    (guint)DDC_DATA_LEN_DFT);
+		return FALSE;
+	}
+
+	/* verify read buffer: overflow guard from the length of data */
+	report_data_sz = buf[1] - DDC_DATA_LEN_DFT;
+	if (report_data_sz + 3 > sizeof(buf)) {
+		g_set_error(error,
+			    G_IO_ERROR,
+			    G_IO_ERROR_INVALID_DATA,
+			    "invalid read buffer: size 0x%02x exceeded 0x%02lx.",
+			    (guint)report_data_sz,
+			    sizeof(buf));
+		return FALSE;
+	}
+
+	/* verify read buffer: match the checksum */
+	checksum ^= DDC_HOST_I2C_ADDR_CHKSUM;
+	for (gsize i = 0; i < report_data_sz + 2; i++)
+		checksum ^= buf[i];
+
+	if (buf[report_data_sz + 2] != checksum) {
+		g_set_error(error,
+			    G_IO_ERROR,
+			    G_IO_ERROR_INVALID_DATA,
+			    "invalid read buffer, checksum expected 0x%02x, got 0x%02x.",
+			    checksum,
+			    buf[report_data_sz + 2]);
+		return FALSE;
+	}
+
+	/* checksum is needn't */
```

Comment:
needed?

Reply:
the `cheksum` is a byte at the tail of `buf`, we don't need it into the byte_array.

Reply:
ohh, i meant more the spelling of needn't -- do you mean `needed`?

Reply:
Ah, okay will refine it in next push.